sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs

```diff
@@ -0,0 +1,114 @@
+            if (userId != default)
+            {
+                var subject = new Claim("sub", userId);
+                claims.Add(subject);
+            }
+            if (roles != default && roles.Length > 0)
+            {
+                var jsonArray = BinaryData.FromObjectAsJson(roles).ToString();
+                var role = new Claim("role", jsonArray);
+                claims.Add(role);
+            }
+
+            var endpoint = _endpoint.ToString().EndsWith("/", StringComparison.Ordinal) ? _endpoint.ToString() : _endpoint + "/";
+            var audience = $"{endpoint}client/hubs/{hub}";
+
+            var token = WebPubSubAuthenticationPolicy.GenerateAccessToken(audience, claims: claims, _credential, expireAfter);
+
+            var clientEndpoint = new UriBuilder(endpoint);
+            clientEndpoint.Scheme = "wss";
+            var uriString = $"{clientEndpoint}client/hubs/{hub}?access_token={token}";
+
+            return new Uri(uriString);
+        }
+
+        /// <summary>
+        /// Parse connection string to endpoint and credential
+        /// </summary>
+        /// <returns></returns>
+        internal static (Uri Endpoint, AzureKeyCredential Credential) ParseConnectionString(string connectionString)
+        {
+            var properties = connectionString.Split(PropertySeparator, StringSplitOptions.RemoveEmptyEntries);
+
+            var dict = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var property in properties)
+            {
+                var kvp = property.Split(KeyValueSeparator, 2);
+                if (kvp.Length != 2)
+                    continue;
+
+                var key = kvp[0].Trim();
+                if (dict.ContainsKey(key))
+                {
+                    throw new ArgumentException($"Duplicate properties found in connection string: {key}.");
+                }
+
+                dict.Add(key, kvp[1].Trim());
+            }
+
+            if (!dict.TryGetValue(EndpointPropertyName, out var endpoint))
+            {
+                throw new ArgumentException($"Required property not found in connection string: {EndpointPropertyName}.");
+            }
+            endpoint = endpoint.TrimEnd('/');
+
+            if (!dict.TryGetValue(AccessKeyPropertyName, out var accessKey))
+            {
+                throw new ArgumentException($"Required property not found in connection string: {AccessKeyPropertyName}.");
+            }
+
+            int? port = null;
+            if (dict.TryGetValue(PortPropertyName, out var rawPort))
+            {
+                if (int.TryParse(rawPort, out var portValue) && portValue > 0 && portValue <= 0xFFFF)
+                {
+                    port = portValue;
+                }
+                else
+                {
+                    throw new ArgumentException($"Invalid Port value: {rawPort}");
+                }
+            }
+
+            var uriBuilder = new UriBuilder(endpoint);
```

Comment:
What does this throw if the `endpoint` was malformed?  Should we wrap it and throw our own `ArgumentException`?

Reply:
I think wrapping like this makes sense for runtime errors, but not really for usage errors. Malformed URI is a usage error. But it is a very good discussion we should have at some point: which exceptions we are ok with leaking and which not, and how to ensure our codebase is clean re this issue.

Reply:
I like to think a lot about how this will go wrong for customers in practice.  They'll build their app, test it, verify everything's good, deploy it, and move on to the next project.  Six months later, somebody decides to roll credentials and updates the KV secret with this connection string but misses a character when copy/pasting but doesn't notice right away.  Suddenly this app that's worked fine forever starts throwing exceptions about Malformed URIs.  How frustrating would that experience be?  And is it worth making that easier by saying "Failed to parse connection string because Endpoint is malformed" (with a more detailed inner exception) so they're that much closer to the real cause of the error?

Reply:
OK. This is a good point. I thought you just wanted to change the exception type, but what you want is to give people a better message. I agree with that. I will add it to the list of things to improve after beta. I worry that this is an example of a much broader set of issues.